### PR TITLE
performance log analysis #8357

### LINKED ISSUE #8357: Detailed performance analysis of getAllAfter and getIndexList methods [5]
### Problem Description

Retrieving data via ReST during the mobile app synchronisation process can take a lot of time (with as few as 1000 cases taking several minutes to synchronize). We've also received reports (and witnessed) the web app directories taking more than a few seconds to load on systems with a lot of data (e.g. more than 500,000 entries in a directory).

### Proposed Solution

In preparation of potential refactorings and code improvements, analyse the performance of the getAllAfter and getIndexList methods of the following entities:

* Case
* Contact
* Sample
* Person

This should include all EJBs called in the process, and also especially include the pseudonymisation logic because this was already identified as a huge bottleneck at least for the synchronisation process.

Step 1: Activate `PerformanceLoggingInterceptor` on TRACE level to analyse call hierarchy and where time is spent. Document execution times.
Step 2: Dive deeper into hotspot method to find costly transformations (probably in `toDto`/`pseudonymize` methods).

### Additional Information

A potential follow-up issue to improve the pseudonymisation process could include the following action points:
- Skip pseudonymisation completely for entites that we know will not be pseudonymised based on the user's user rights (e.g. an entity inside the user's jurisdiction does not need to be checked for pseudonymisation if the user has both the SEE_PERSONAL_DATA_IN_JURISDICTION and SEE_SENSITIVE_DATA_IN_JURISDICTION user rights)
- Create a cache per entity of all fields that need to be considered during pseudonymisation in order to reduce the usage of reflection

Since the toDto logic also takes some time, another possible change could be to directly select into DTOs, even though that would potentially introduce a significantly increased maintenance effort.

As database where we want to reproduce  the perf. issues and measure the improvements we are using the pseudonymized data from France.

### Deliverables
Confirm that pseudonymization is one cause 
Explore what else causes performance issues
Document KPI (Key performance indicators) and preconditions to be able to measure the improvement
Create follow-up tickets that are obvious

==>
- pseudonymization can be ruled out as a relevant impact on performance
- long runtimes are mainly caused by long running SQL queries with massive joins
- follow-up issues: #8938 for `getIndexList` methods, #8946 for `getAll` methods in the REST API

### PR DESCRIPTION
**Pull request by syntakker**
This is not the final fix for #8357, please make sure not to close the issue when reviewing!

Fixes #8357

### CODE REVIEW

**Comment by StefanKock** [comment]
@MateStrysewske please look into this and help to provide an extension to the troubleshooting guide.

File: sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+            "</html>\n";
+    //@formatter:on
+
+	public static final int TIME_TRESHOLD_YELLOW = 100;
+	public static final int TIME_TRESHOLD_ORANGE = 300;
+	public static final int TIME_TRESHOLD_RED = 1000;
+
+	public static final String OUTPUT_DIRECTORY = "target/performance/";
+
+	private Map<String, Stack<String>> callstacks;
```
Comment (StefanKock) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java:70]: Why instance variable? Better move to method if not needed outside

File: sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+	private Map<String, MethodStats> methodStats;
```
Comment (StefanKock) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java:71]: Why instance variable? Better move to method if not needed outside

File: sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+import java.util.regex.Pattern;
+
+import org.apache.commons.io.FilenameUtils;
+import org.jetbrains.annotations.NotNull;
+
+public class PerformanceLogAnalysisGenerator {
+
+	private static final Pattern startPattern =
+		Pattern.compile("([^ ]* [^ ]*) TRACE *(.*) d.s.s.b.u.PerformanceLoggingInterceptor - Started: ([^ ]*) with parameters .*$");
+	private static final Pattern finishPattern =
```
Comment (StefanKock) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java:40]: Contant -> uppercase

**Review by StefanKock — CHANGES_REQUESTED** [review]

File: sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java (hunk 1 above)
Comment (syntakker) [reply sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java:90]: Now used outside the method in `getUnfinishedMethodsHtml`.

File: sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java (hunk 2 above)
Comment (syntakker) [reply sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java]: Obsolete and removed.

File: sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java (hunk 3 above)
Comment (syntakker) [reply sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java]: done.

File: sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+import java.util.Stack;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.io.FilenameUtils;
+import org.jetbrains.annotations.NotNull;
+
+public class PerformanceLogAnalysisGenerator {
+
+	private static final Pattern startPattern =
```
[the remark this answers is not in the document]
Comment (syntakker) [reply sormas-backend/src/test/java/de/symeda/sormas/backend/performance/PerformanceLogAnalysisGenerator.java]: done.

**Comment by syntakker** [comment]
Description added in `TROUBLESHOOTING.md`.

**Review by StefanKock — APPROVED** [review]

**Pull request merged**